File: frontend/src/components/pages/automation/mindmaps/mindMapDocument.ts
// TS mirror of the backend MindMapDocument JSON contract (camelCase).
// All helpers are pure: they return a NEW document and never mutate inputs.

export type MindMapNodeStatus = "active" | "done" | "blocked" | "idea";

export interface MindMapNodePosition {
  x: number;
  y: number;
}

export interface MindMapNode {
  id: string;
  parentId: string | null;
  title: string;
  notes: string | null;
  status: MindMapNodeStatus;
  owner: string | null;
  lockedBy: string | null;
  sourceMeetingIds: string[];
  position: MindMapNodePosition | null;
  collapsed: boolean;
}

export interface SuppressedNode {
  title: string;
  deletedBy: string | null;
}

export interface MindMapDocument {
  schemaVersion: number;
  rootNodeId: string;
  nodes: MindMapNode[];
  suppressedNodes: SuppressedNode[];
}

/** The subset of a node the user can edit by hand. */
export type MindMapNodePatch = Partial<Pick<MindMapNode, "title" | "notes" | "owner" | "status">>;

export function parseDocument(json: string): MindMapDocument {
  const parsed = JSON.parse(json) as MindMapDocument;
  if (!parsed || !Array.isArray(parsed.nodes) || !parsed.rootNodeId) {
    throw new Error("Invalid mind map document");
  }
  // toMindElixir throws if rootNodeId names no node, and it is called unguarded
  // from MindMapCanvas's mount and revision effects. Catching the same defect here
  // — where every caller already has a try/catch and an error state to fall back
  // on — keeps a malformed document from escaping as an uncaught exception later.
  if (!parsed.nodes.some((n) => n.id === parsed.rootNodeId)) {
    throw new Error("Invalid mind map document: no node matches rootNodeId");
  }
  return parsed;
}

File: backend/src/Anela.Heblo.Application/Features/MindMaps/Model/MindMapDocument.cs
namespace Anela.Heblo.Application.Features.MindMaps.Model;

/// <summary>
/// The whole mind map document persisted as one jsonb value (camelCase on the wire).
/// Position/Collapsed/LockedBy are UI/system metadata the LLM never writes —
/// MindMapGuard restores them from the previous version after every LLM update.
/// </summary>
public class MindMapDocument
{
    public int SchemaVersion { get; set; } = 1;
    public string RootNodeId { get; set; } = null!;
    public List<MindMapNode> Nodes { get; set; } = new();
    public List<SuppressedNode> SuppressedNodes { get; set; } = new();
}

public class MindMapNode
{
    public string Id { get; set; } = null!;
    public string? ParentId { get; set; }
    public string Title { get; set; } = null!;
    public string? Notes { get; set; }
    public string Status { get; set; } = MindMapNodeStatus.Active;
    public string? Owner { get; set; }
    public string? LockedBy { get; set; }
    public List<Guid> SourceMeetingIds { get; set; } = new();
    public NodePosition? Position { get; set; }
    public bool Collapsed { get; set; }
}

public class NodePosition
{
    public double X { get; set; }
    public double Y { get; set; }
}

public class SuppressedNode
{
    public string Title { get; set; } = null!;
    public string? DeletedBy { get; set; }
}

public static class MindMapNodeStatus
{
    public const string Active = "active";
    public const string Done = "done";
    public const string Blocked = "blocked";
    public const string Idea = "idea";

    public static readonly IReadOnlyList<string> All = new[] { Active, Done, Blocked, Idea };
}
